Adding ml-based techniques to SEMAPV

matentzn opened the pull request:
Fixes #27, #28.

cc @sven-h @cassiatrojahn

## What's in

Seven new classes covering the matching techniques that have shown up in OAEI and elsewhere over the last few years but weren't yet in SEMAPV, plus `BoundedPathMatching` from #28:

```
semapv:Matching
├── semapv:BoundedPathMatching              ⊑ semapv:StructuralMatching
├── semapv:MachineLearningBasedMatching
│   ├── semapv:EmbeddingBasedMatching
│   │   ├── semapv:TransformerBasedMatching
│   │   └── semapv:GraphRepresentationLearningBasedMatching
│   └── semapv:LLMBasedMatching
└── semapv:AgentBasedMatching
```

## Why this shape

The original proposal in #27 had transformer, LLM, ML and graph-RL matching all as flat siblings under `semapv:Matching`. That loses the actual taxonomy and breaks queries like "all ML-based matching processes". The hierarchy above factors them on two orthogonal axes:

- **What kind of model** — captured by `MachineLearningBasedMatching` as a parent.
- **How the model is used** — embeddings compared by similarity vs. a model making the decision directly. That distinction is what separates `EmbeddingBasedMatching` from `LLMBasedMatching`, not the architecture (transformer vs. anything else). So `TransformerBasedMatching` is specifically "transformer-derived embeddings for similarity matching" (BERTMap, sentence-BERT, OpenAI embeddings), and `LLMBasedMatching` is the LLM-as-judge / fine-tuned-generator flavour.

`AgentBasedMatching` is a direct child of `semapv:Matching` rather than of `LLMBasedMatching`, on purpose. The distinguishing feature of agent-based matching is the workflow shape — an autonomous controller that decides at runtime which tools to call, what to retrieve, and when to stop — not the underlying model. Today most agent-based matchers happen to be LLM-driven, so a concrete process will often be classified as both `AgentBasedMatching` and `LLMBasedMatching`; the two are orthogonal, not disjoint. Subclassing under `LLMBasedMatching` would have been an OWL universal claim that every agentic matcher is LLM-based, which is true today but brittle (symbolic planners, RL agents, neurosymbolic agents, whatever comes after transformers).

## Definitions

Definitions describe the matching approach — not the technology — to stay consistent with the rest of the vocabulary (`BackgroundKnowledgeBasedMatching`, `InstanceBasedMatching`, etc.). Architectural background goes into `rdfs:comment` instead, so a reader who already knows what a transformer or an LLM is doesn't have to wade through textbook prose to find out what the matching process does.

Overlap notes are baked into `rdfs:comment` where they matter:

- `LLMBasedMatching` notes that using an LLM only to produce embeddings is not LLM-based matching but transformer-based — the classifier is *how* the model is used, not *which* model.
- `LLMBasedMatching` also points to `AgentBasedMatching` for the multi-step / autonomous case.
- `AgentBasedMatching` notes the distinction from both single-step LLM matching and composite matching (the latter is multi-technique under a fixed pipeline; agent-based is runtime-controlled).

## Other changes

- Small typo fix in `BlankNormalisation` example (extra space inside the quoted "after" string).
- `dc:source` left empty on the new classes for now — happy to add references in a follow-up once we've agreed which OAEI / Matcha / BERTMap papers to anchor to.

matentzn committed:
Add ML-based techniques to semapv

matentzn committed:
Update documentation

matentzn committed:
Merge branch 'main' into addingmltechniques

matentzn committed:
Modernise semapv PR with ml based techniques

cthoyt commented on semapv-terms.tsv:
```diff
@@ -47,3 +47,10 @@ semapv:TokenBasedDistance	token-based distance		https://doi.org/10.1007/978-3-64
 semapv:LanguageBasedSimilarityMeasure	language-based similarity measure		https://doi.org/10.1007/978-3-642-38721-0				semapv:SimilarityMeasure
 semapv:Tokenization	tokenization		https://doi.org/10.1007/978-3-642-38721-0				semapv:Normalization
 semapv:TermExtraction	term extraction		https://doi.org/10.1007/978-3-642-38721-0			named entity recognition	semapv:Normalization
+semapv:BoundedPathMatching	bounded path matching	A structural matching process based on the comparison of matched super- and subclasses (paths) of two entities.					semapv:StructuralMatching
```
not clear what this is. can you please give a concrete example of a piece of software that does this, and of a specific mapping?

cthoyt commented on semapv-terms.tsv:
```diff
@@ -47,3 +47,10 @@ semapv:TokenBasedDistance	token-based distance		https://doi.org/10.1007/978-3-64
+semapv:MachineLearningBasedMatching	machine learning-based matching process	A matching process in which matches between entities are predicted by machine learning models trained on labelled data or learned representations.		Matches between entities are produced by applying a trained model to features derived from entity descriptions or their surrounding context.		ML-based matching	semapv:Matching
+semapv:EmbeddingBasedMatching	embedding-based matching process	A machine learning-based matching process in which matches are identified by comparing learned vector representations (embeddings) of entities.		Matches between entities are identified by comparing their embedded representations under a vector similarity measure such as cosine similarity.		embedding similarity-based matching	semapv:Ma ...
+semapv:TransformerBasedMatching	transformer-based matching process	An embedding-based matching process that uses transformer models to produce contextual vector representations of entity features for similarity comparison.		Matches between entities are established by comparing contextual representations of entity features produced by a transformer model.	Transformer models are a deep learning arc ...
+semapv:GraphRepresentationLearningBasedMatching	graph representation learning-based matching process	An embedding-based matching process in which matches are identified by comparing learned vector representations that encode the structural and semantic features of entities and their surrounding graph context.		Matches between entities are identified by analyzing learned representations that encod ...
```
This is usually called `graph machine learning`. I've never heard the phrase `graph representation learning` before

cthoyt reviewed (changes requested):
I would like to see concrete references to example software and/or publications where each of the proposed matching processes have been implemented and/or described

matentzn committed:
Add Blocking subtree, dc:source citations, and concrete examples

Adds semapv:Blocking and semapv:EmbeddingBasedBlocking to capture the
OAEI candidate-generation step separately from matching. Every new ML
class now has a dc:source DOI and a concrete software example (BERTMap,
OLaLa, OntoAligner, sentence-BERT, FAISS, node2vec). BoundedPathMatching
gets a concrete example and the Euzenat & Shvaiko citation. Added
"graph machine learning-based matching" as altLabel on the graph
representation learning class.

matentzn committed:
MISC fixes

cthoyt commented on semapv-terms.tsv:
```diff
@@ -47,3 +47,12 @@ semapv:TokenBasedDistance	token-based distance		https://doi.org/10.1007/978-3-64
+semapv:BoundedPathMatching	bounded path matching	A structural matching process based on the comparison of the matched super- and subclass paths (up to a bounded depth) of two candidate entities.	https://doi.org/10.1007/978-3-642-38721-0	Two classes are matched because their direct superclasses and subclasses (within a bounded depth) are already paired by other matchers, and the structural agreeme ...
+semapv:MachineLearningBasedMatching	machine learning-based matching process	A matching process in which matches between entities are predicted by machine learning models trained on labelled data or learned representations.	https://doi.org/10.1145/511446.511532	A trained machine learning model classifies a pair of candidate entities as matching or not based on learned features; the OntoAligner too ...
+semapv:EmbeddingBasedMatching	embedding-based matching process	A machine learning-based matching process in which matches are identified by comparing learned vector representations (embeddings) of entities under a vector similarity measure.	https://doi.org/10.48550/arXiv.2310.11244	A sentence-transformer encodes the label of each candidate entity; the pair is reported as a match when the cosine s ...
+semapv:TransformerBasedMatching	transformer-based matching process	An embedding-based matching process that uses transformer models to produce contextual vector representations of entity features for similarity comparison.	https://doi.org/10.48550/arXiv.2112.02682	BERTMap fine-tunes BERT on synonym pairs derived from the input ontologies and matches classes whose fine-tuned contextual embeddings  ...
```
not a blocking issue, but we are now going to create a shadow concept for the hierarchy of different kinds of ML models here. maybe would be nice if there's a high quality ontology of ML model types to point to. Save this idea for later ;)

matentzn replied:
In principle I agree, but as the blocking discussion with @sven-h  showed, a "Matching" process does more than identify a ML model: it identifies a model and matching decision point (even if its sometimes left a bit vague how the decision is made). Furthermore, there are only a few "general approaches" that lend themselves to matching, so hopefully that will keep the hierarchy tight. 

In principle I agree with the risk of course.

cthoyt reviewed (approved):
looks good! let's merge this and nitpick the descriptions in follow-ups
